### DISCUSSION TITLE
[V5] Namespaces socket.io with V5

### DISCUSSION THREAD

**Discussion by Nikolastual** [adonisjs/core#1878] [Help]
Hi ! 

Reading the doc : [Using Socket.io with AdonisJS v5](https://preview.adonisjs.com/blog/socketio-with-adonisjs-v5) 
I don't know how to create namespaces with the `start()` function in Ws.ts : 

```
import socketIo from 'socket.io'
import Server from '@ioc:Adonis/Core/Server'

class Ws {
	public isReady = false
	public io: socketIo.Server

	public start (callback: (socket: socketIo.Socket) => void) {
		this.io = socketIo(Server.instance!, { path: '/ws', transports: ['websocket', 'polling'] })
		this.io.on('connection', callback)
		this.isReady = true
	}
}

/**
 * This makes our service a singleton
 */
export default new Ws()
```
My file socket.ts : 

```
import Ws from 'App/Services/Ws'

Ws.start((socket) => {
      console.log(socket.id)
})
```

**Comment by DraftProducts** [comment]
Hi, @FrenchNicky you have to create namespaces from your server instance.
With Socket.io you have two notions, rooms and namespaces (nsp).
 - Namespaces can be create from `Server.of('/namespace')`
 - Rooms can be create from `Socket.to('/room_id')`

So if you need a namespace you can create it from `this.io`.
You can also provide your io server to your start callback: 
```ts
// Ws
public start (callback: (socket: socketIo.Socket) => void) {
    this.io = socketIo(Server.instance!, { path: '/ws', transports: ['websocket', 'polling'] })
    this.io.on('connection', (socket) => callback(this.io, socket))
    this.isReady = true
}

// socket.ts
Ws.start((io, socket) => {
    console.log(socket.id)
    io.of('namespace')
})
```

**Reply by DraftProducts** [reply]
I'm currently working on a project with Socket.io, so if you need an example here's the socket.io part.
https://github.com/Bio2Game/Bio2Game-API/tree/master/socket.io

**Reply by Nikolastual** [reply]
Thanks !

But when I do an emit on the client side : `msg.emit('test', 'Test emit')` the logs are doubled in the console with each page refresh : 

![Capture d’écran 2020-11-08 à 20 52 23](https://user-images.githubusercontent.com/43614012/98482842-67b41200-2204-11eb-8441-6870f00f61bf.png)

```javascript
import Ws from 'App/Services/Ws'

Ws.start((io, socket) => {
	let messages = io.of('/messages')
	messages.on('connection', (socket) => {
		console.log('connection',socket.id)

		socket.on('test', (data:any) => {
			console.log(data)
		})

		socket.on('disconnect', () => {
			console.log('disconnected', socket.id)
		})

	})
})
```

If I want to use namespaces, I can't use Ws.ts file:

it works like this :
```javascript
// socket.ts
import SocketIO from 'socket.io'
import Server from '@ioc:Adonis/Core/Server'

const io = SocketIO(Server.instance!, { path: '/ws', transports: ['websocket', 'polling'] })
const msg = io.of('/messages'), alerts = io.of('/notifications')

...
```

I will work later to gather my code in class ;)

**Comment by rbartholomay** [comment]
Hi there!

How can use the adonisjs auth inside this socket.io implementation?
